proof-of-spacetime.md

```diff
@@ -0,0 +1,198 @@
+
+If the miner knows that the sectors are permanently lost, they can submit them as part of the `doneSet`, to ensure they are removed from the proving set.
+
+
+{{% notice todo %}}
+**TODO**: The penalization for faults is not clear yet.
+{{% /notice %}}
+
+#### Generation
+
+`GeneratePoSt` generates a __*Proof of Spacetime*__ over all  __*sealed sectors*__ of a single miner— identified by their `commR` commitments. This is accomplished by performing a series of merkle inclusion proofs (__*Proofs of Retrievability*__). Each proof is of a challenged node in a challenged sector. The challenges are generated pseudo-randomly, based on the provided `seed`. At each time step, a number of __*Proofs of Retrievability*__ are performed.
+
+```go
+// Generate a new PoSt.
+func GeneratePoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, faults FaultSet) PoStProof {
+    // Generate the Merkle Inclusion Proofs + Faults
+
+    inclusionProofs := []
+	sectorsSorted := []
+    challenges := DerivePoStChallenges(seed, faults, sectorSize, len(sectors))
+
+    for challenge in challenges {
+        // Leaf index of the selected sector
+        inclusionProof, isFault := GenerateMerkleInclusionProof(challenge.Sector, challenge.Leaf)
+        if isFault {
+            // faulty sector, need to post a fault to the chain and try to recover from it
+            return Fatal("Detected late fault")
+        }
+
+        inclusionProofs[n] = inclusionProof
+		sectorsSorted[i] = sectors[sector]
+    }
+
+    // Generate the snark
+    snark_proof := GeneratePoStSnark(sectorSize, challenges, sectorsSorted, inclusionProofs)
+
+    proof := PoStProof {
+        snark: snark_proof
+    }
+
+    return proof, faults
+}
+```
+
+#### Verification
+
+`VerifyPoSt` is the functional counterpart to `GeneratePoSt`. It takes all of `GeneratePoSt`'s output, along with those of `GeneratePost`'s inputs required to identify the claimed proof. All inputs are required because verification requires sufficient context to determine not only that a proof is valid but also that the proof indeed corresponds to what it purports to prove.
+
+```go
+// Verify a PoSt.
+func VerifyPoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, proof PoStProof, faults FaultSet) bool {
+    challenges := DerivePoStChallenges(seed, faults, sectorSize, len(sectors))
+    sectorsSorted := []
+
+    // Match up commitments with challenges
+    for challenge in challenges {
+        sectorsSorted[i] = sectors[challenge.Sector]
+    }
+
+    // Verify snark
+    VerifyPoStSnark(sectorSize, challenges, sectorsSorted)
+}
+```
+
+
+#### Types
+
+```go
+// The final proof stored on chain.
+type PoStProof struct {
+    snark []byte
+}
+```
+
+```go
+// The random challenge seed, provided by the chain.
+Seed [32]byte
+```
+
+```go
+type Challenge struct {
+    Sector Uint
+    Leaf Uint
+}
+```
+
+#### Challenge Derivation
+
+```go
```

Comment:
The code block is marked as `go`, but the code inside it is a mixture of Go and Rust (e.g. `for n in 0..POST_CHALLENGES_COUNT`).